In [68]:
import numpy as np
import pandas as pd

#task 1
df=pd.read_csv("adult.csv")
df.info()
df.head(10)
print(df.isnull().sum())
print(df.replace("?",np.nan,inplace=True))

#task 2
print("task 2")
print("descending order")
print(df.isnull().sum().sort_values(ascending=False))
print("sort")
#print(df.sort_values(ascending=False))
#print(df.isnull().sum().sort(reverse=True))

#task 3
print("task 3")
print("for workclass")
print(df['workclass'].replace("?","unknown",inplace=False).fillna("unknown"))
print("for occupation")
print(df['occupation'].replace("?","unknown",inplace=False).fillna("unknown"))
print("we have chosen unknown because taking mode of a string is not feasible and dropping them would reduce the sample size hence we replaced it with unknown")
print("there is no numeric column having a ? ")


#task 4
print("task 4")
exact=df.duplicated().sum()
print("this is the number of rows that are exactly duplicate =", exact)
feature_cols=df.columns.drop("age")
feature_duplicates=df.duplicated(subset=feature_cols).sum()
print("this is the number of duplications without any label= ",feature_duplicates)
print("these should not be counted as duplication because having the same feature or labels does not mean that both entities will yield same results hence they represent distinct obs")


#task 5
print("task 5")
print("these are the unique values in education column= ",df['education'].str.strip().unique())
print("these are the unique values in marital status column= ",df['marital.status'].str.strip().unique())
print("these are the unique values in native country column= ",df['native.country'].str.strip().unique())
print("they might get categorized as diff strings")

#task 6
print("task 6")
df.describe()
print(df.describe(include='object'))

#task 7
print("task 7")
most_common_occupation = df['occupation'].value_counts().idxmax()
print("Most common occupation:", most_common_occupation)

sex_percentage = df['sex'].value_counts(normalize=True) * 100
print("Percentage distribution of sex:")
print(sex_percentage)

income_percentage = df['income'].value_counts(normalize=True) * 100
print("Percentage distribution of income:")
print(income_percentage)

#task 8
print("task 8")
print(df.groupby('education')['education.num'].unique())
inconsistencies = df.groupby('education')['education.num'].unique()
inconsistencies[inconsistencies.apply(len) > 1]
print("there are no inconsisties because each education entity maps to a unique education number")
task 9
print("task 9")
import pandas as pd
import sqlite3

df = pd.read_csv("adult.csv")
conn = sqlite3.connect("adult.db")
df.to_sql("adult_income", conn, if_exists="replace", index=False)

result = pd.read_sql_query(
    "SELECT * FROM adult_income WHERE age > 30", conn
)

print("Original shape:", df.shape)
print("Extracted shape:", result.shape)

conn.close()

#task 9
print("task 9")
import pandas as pd
import sqlite3

df = pd.read_csv("adult.csv")
conn = sqlite3.connect("adult.db")
df.to_sql("adult_income", conn, if_exists="replace", index=False)

result = pd.read_sql_query(
    "SELECT * FROM adult_income WHERE age > 30", conn
)

print("Original shape:", df.shape)
print("Extracted shape:", result.shape)

conn.close()

#task 10
print("task 10")








<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education.num   32561 non-null  int64 
 5   marital.status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital.gain    32561 non-null  int64 
 11  capital.loss    32561 non-null  int64 
 12  hours.per.week  32561 non-null  int64 
 13  native.country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB
age               0
workclass         0
fnlwgt            0
education         0
education.num

the dataset contains 32,561 rows and 15 columns. It includes demographic and employment information about individuals, such as age, education, occupation, workclass, marital status, working hours, capital gains, and income.

The target variable is income, which has two classes:
- <=50K
- >50K


### Missing Values
Missing values were represented as **"?"** in some categorical columns.

The columns with the highest missing values were:
- Occupation: 1,843 missing values
- Workclass: 1,836 missing values
- Native country: 583 missing values

### Duplicate Records
The dataset contained:
- 24 exact duplicate rows
- 67 duplicate rows without considering the income label

Duplicates were not removed because the same feature values can belong to different individuals.

### Data Consistency
- No issues were found between **education** and **education.num**.
- All numeric columns were complete and valid.


### Income Distribution
The dataset is imbalanced:
- <=50K: 75.92%
- >50K: 24.08%

### Gender Distribution
- Male: 66.92%
- Female: 33.08%

### Common Categories
The most frequent categories were:

- Workclass: Private
- Education: HS-grad
- Marital Status: Married-civ-spouse
- Race: White
- Native Country: United-States

The most common occupation was **Prof-specialty**.

## Data Cleaning Steps

The following preprocessing steps were applied:

- Replaced missing values marked as "?" with **"unknown"**.
- Did not remove rows with missing values to avoid losing data.
- No numeric imputation was needed because numeric columns had no missing values.
- Checked duplicate records but kept them because they may represent different individuals.
- Verified that categorical values were consistent.

After cleaning, the final dataset contained:

**21,989 rows and 15 columns**